pkg/agent/controller/noderoute/node_route_controller.go

```diff
@@ -144,35 +147,178 @@ func (c *Controller) enqueueNode(obj interface{}) {
+	// cluster.
+	klog.V(4).Infof("Need to delete %d routes", len(routes))
+	for _, route := range routes {
+		if err := netlink.RouteDel(route); err != nil {
+			klog.Errorf("error when deleting route '%v': %v", route, err)
+		}
+	}
+
+	return nil
+}
+
+// removeStaleTunnelPorts removes all the tunnel ports which no longer correspond to a Node in the
+// cluster. If the antrea agent restarts and Nodes have left the cluster, this function will take
+// care of removing tunnel ports which are no longer valid. If the tunnel port configuration has
+// changed, the tunnel port will also be deleted (the controller loop will later take care of
+// re-creating the port with the correct configuration).
+func (c *Controller) removeStaleTunnelPorts() error {
+	nodes, err := c.nodeLister.List(labels.Everything())
+	if err != nil {
+		return fmt.Errorf("error when listing Nodes: %v", err)
+	}
+	// desiredInterfaces is the set of interfaces we wish to have, based on the current list of
+	// Nodes. If a tunnel port corresponds to a valid Node but its configuration is wrong, we
+	// will not include it in the set.
+	desiredInterfaces := make(map[string]bool)
+	// knownInterfaces is the list of interfaces currently in the local cache.
+	knownInterfaces := c.interfaceStore.GetInterfaceKeysByType(interfacestore.TunnelInterface)
+
+	if c.ipsecPSK != "" {
+		for _, node := range nodes {
+			interfaceConfig, found := c.interfaceStore.GetNodeTunnelInterface(node.Name)
+			if !found {
+				// Tunnel port not created for this Node, nothing to do.
+				continue
+			}
+
+			peerNodeIP, err := getNodeAddr(node)
+			if err != nil {
+				klog.Errorf("Failed to retrieve IP address of Node %s: %v", node.Name, err)
+				continue
+			}
+
+			portName := util.GenerateNodeTunnelInterfaceName(node.Name)
+			validConfiguration := interfaceConfig.PSK == c.ipsecPSK &&
+				interfaceConfig.RemoteIP.Equal(peerNodeIP) &&
+				interfaceConfig.TunnelInterfaceConfig.Type == c.tunnelType
+			if validConfiguration {
+				desiredInterfaces[portName] = true
+			}
+		}
+	}
+
+	// remove all ports which are no longer needed or for which the configuration is no longer
+	// valid.
+	for _, ifaceID := range knownInterfaces {
+		if _, found := desiredInterfaces[ifaceID]; found {
+			// this interface matches an existing Node, nothing to do.
+			continue
+		}
+		interfaceConfig, found := c.interfaceStore.GetInterface(ifaceID)
+		if !found {
+			// should not happen, nothing should have concurrent access to the interface
+			// store for tunnel interfaces.
+			klog.Errorf("Interface %s can no longer be found in the interface store", ifaceID)
+			continue
+		}
+		if interfaceConfig.OFPort == types.DefaultTunOFPort {
+			continue
+		}
+		if err := c.ovsBridgeClient.DeletePort(interfaceConfig.PortUUID); err != nil {
+			klog.Errorf("Failed to delete OVS tunnel port %s: %v", interfaceConfig.InterfaceName, err)
+		}
+		c.interfaceStore.DeleteInterface(interfaceConfig)
```

Comment:
If deleting OVS port failed, better to keep the interface in cache?

Reply:
Humm.. then we might still need to handle the tunnel type/Node change in createIPSecTunnelPort(). Maybe keep the TODO there?